### ISSUE TITLE
Types not checked when computed property key has multiple possible values

### ISSUE THREAD

**Issue by mabeyj** [microsoft/TypeScript#32309]
**TypeScript Version:**  3.5.2 and 3.6.0-dev.20190709

**Search Terms:** computed property key name type

**Code**

```ts
interface IProps {
  a?: string;
  b?: number;
}

function test1a(key: string, value: string): IProps {
  if (key === "a" || key === "b") {
    return {[key]: value}; // Should error
  }
  return {};
}

function test1b(key: string, value: string): IProps {
  if (key === "a") {
    return {[key]: value};
  }
  if (key === "b") {
    return {[key]: value}; // Error as expected
  }
  return {};
}

function test1c(key: string, value: string): IProps {
  const props: IProps = {};
  if (key === "a" || key === "b") {
    props[key] = value; // Error as expected
  }
  return props;
}

function test2a(key: string, value: string): IProps {
  return {[key]: value}; // Should error
}

function test2b(key: string, value: string): IProps {
  const props: IProps = {};
  props[key] = value; // Error as expected
  return props;
}
```

**Expected behavior:**

* `test1a` should raise an error because:
  * The only possible computed keys in this case are `a` and `b`. The types of both properties `a` and `b` are not satisfied.
  * `test1b` is logically the same as `test1a`, yet it does raise an error. The difference being there is only one possible computed key in each `if` block instead of multiple possible keys.
  * `test1c` uses assignment instead of computed properties, and it does raise an error in this case.
* `test2a` should cause an error because:
  * The interface has no index signature.
  * `test2b` uses assignment instead of computed properties, and it does raise an error about the missing index signature in this case.

**Actual behavior:**

* `test1a` and `test2a` don't raise any errors, thus allowing objects to be returned that don't satisfy the interface.

**Playground Link:** 

https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgJIAUoHsAOBnZAbwFgAoZZOAfgC5k8wpQBzAbjIoCNbkQBXALado7UgF8yZGHxAIwwLCGSQGARjgAKANYQAnnQZMQzADTIAbnAA2fCAcYsAlHQzZ8RDsmAxk2vcgBeIOQAIjgQ5AAfSOQdXUDgkM4Qxw9yCmQoCDA+KCVCAG04gF06SxsIMVZkAHoa5ABlAAssPisAE2RobChPCXSsnLyiKrJ+qRk5BSUVMFVOP316B2Mzctt7I2ZnNExcAhJ07184hIDQ8NTDjMzs3PyivVKLa1tR9P6KY8Wz0OSrzwUQb3IiPXTPdaVap1ZAAUSgPUoBAgAA8cBA5BB2n1PMDhoR3uNSNJZPJFMoIGoEItNiw1q87MstjtXPs0hQEIoGMgcG48C49u5zgTRF8fD8gucwhForF-JK-il2RlefswcVAi8KtD6vDEXBkWiMZBsR9cXdhqr8KIiSSpuTZgAmTRxWmrLUbJlOAV85V4h4lMoMqq1erNVodLoIrC9cSSYmTMkzSlgR0LV1e92Qt3bH1s67ITkgblW-m7X3C94UUvqzWQnVw6NQJFdI2Y01Ai1KUs2oA

**Related Issues:**

**Comment by RyanCavanaugh** [comment]
The root cause in `test1a` is that the type system isn't able to represent types like this -- there's no concept of a key whose name is one from a set of possible values. So in `test1a`, the type of the expression `{[key]: value}` is the reduced form `{[x: string]: string}`, which is assignable to `IProps` because all of `IProps` members are optional.

That's also what's happening in `test2a`; the assignability rules in general assume that an index signature only contributes positively toward the assignability of a source type. This in general is the assumption that people want, though it certainly isn't sound.

The current rules around index signatures are a best-effort compromise and there isn't any change we could do here that wouldn't break a lot of existing code, unfortunately.

**Comment by fatcerberus** [comment]
> the type system isn't able to represent types like this -- there's no concept of a key whose name is one from a set of possible values

Would the change here https://github.com/microsoft/TypeScript/pull/26797 - if merged - be a meaningful step towards being able to represent such types?
